### PR TITLE
Fix instable `pugClassNotation` with `printWidth`

### LINKED ISSUE #317: Bug: `pugClassNotation` does not respect `printWidth` and is unstable
### Plugin Version

v1.19.0 (not released yet)

### Prettier Version

v2.4.1

### Which frameworks are affected?

- [X] none
- [X] vue
- [ ] angular
- [ ] svelte

### Node Version

v16.13

### Which operating systems have you used?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Prettier config

```js
// @ts-check
/// <reference types="@prettier/plugin-pug/src/prettier" />

/**
 * @type {import('prettier').Options}
 */
module.exports = {
  plugins: [require.resolve('@prettier/plugin-pug')],

  arrowParens: 'always',
  bracketSpacing: true,
  printWidth: 120,
  semi: true,
  singleQuote: true,
  trailingComma: 'none',
  tabWidth: 2,
  useTabs: false,

  pugSingleQuote: false,
  pugFramework: 'vue',
  pugSortAttributesBeginning: [
    '^cols$',
    '^v-for$',
    '^:key$',
    '^v-if$',
    '^v-else-if$',
    '^v-else$',
    '^keypath$',
    '^v-on$',
    '^v-bind$',
    '^ref$',
    '^v-model',
    '^name$',
    '^:?type$',
    '^:value$',
    '^v-text$',
    '^v-t$',
    '^:?label$',
    '^:headers$',
    '^:items$',
    '^:?item-key$',
    '^:?item-text$',
    '^:?item-value$',
    '^:?item-disabled$',
    '^:?must-sort$',
    '^:?sort-by$',
    '^:?sort-desc$',
    '^:?placeholder$',
    '^:?persistent-placeholder$',
    '^:?src$',
    '^:?app$',
    '^:?color$',
    '^:?text-color$',
    '^:?background-color$',
    '^:?dark$',
    '^:?icon$',
    '^:?align$',
    '^:?justify$',
    '^:?dense$',
    '^:?small$',
    '^:?outlined$'
  ],
  pugSortAttributesEnd: [
    '^:?hint$',
    '^:?persistent-hint$',
    '^prepend-',
    '^@click:prepend',
    '^append-',
    '^@click:append',
    '^:to$',
    '^exact$',
    '^:(?!(width|height|loading|disabled|data-))',
    '^target$',
    '^:?width$',
    '^:?height$',
    '^@click',
    '^@',
    '^v-shortkey',
    '^@shortkey',
    '^:loading$',
    '^:disabled$',
    '^:?data-'
  ],

  pugClassLocation: 'after-attributes'
};
```

### Input

```pug
<template lang="pug">
v-row(no-gutters)
  v-col(cols="12")
    .text-body-2.font-weight-medium(v-t="'global.copy-variables-title'")
  v-col(cols="12")
    a.text-body-2.d-inline.mr-4(
      v-for="variable in variables",
      :key="variable",
      v-text="`$\\{${variable}}`",
      @click="copyToClipboard(variable)"
    )
</template>

<script lang="ts" src="./CopyVariables.ts"></script>
```

### Output or Error

Toggles between (`formatOnSave`):

```vue
<template lang="pug">
v-row(no-gutters)
  v-col(cols="12")
    div(v-t="'global.copy-variables-title'").text-body-2.font-weight-medium
  v-col(cols="12")
    a(v-for="variable in variables", :key="variable", v-text="`$\\{${variable}}`", @click="copyToClipboard(variable)").text-body-2.d-inline.mr-4
</template>

<script lang="ts" src="./CopyVariables.ts"></script>
```

And:

```vue
<template lang="pug">
v-row(no-gutters)
  v-col(cols="12")
    div.text-body-2(v-t="'global.copy-variables-title'")
  v-col(cols="12").font-weight-medium
    a.text-body-2(v-for="variable in variables", :key="variable", v-text="`$\\{${variable}}`", @click="copyToClipboard(variable)")
</template>

<script lang="ts" src="./CopyVariables.ts"></script>
```

Also see that `.font-weight-medium` was moved to wrong tag (`v-col`)

### Expected Output

```vue
<template lang="pug">
v-row(no-gutters)
  v-col(cols="12")
    div(v-t="'global.copy-variables-title'").text-body-2.font-weight-medium
  v-col(cols="12")
    a(
      v-for="variable in variables",
      :key="variable",
      v-text="`$\\{${variable}}`",
      @click="copyToClipboard(variable)"
    ).text-body-2
</template>

<script lang="ts" src="./CopyVariables.ts"></script>
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by shadowgate15**
closes #317

### CODE REVIEW

**Review by Shinigami92 — CHANGES_REQUESTED** [review]
Tested it against my codebase and it still toggles the classes back and forth.

You could create a simple example project locally and then use the formatter twice.
If you already have a project, then use something like this in your `package.json` 

`"@prettier/plugin-pug": "link:../../../OpenSource/Shinigami92/plugin-pug"`
Needs to be the relative path to your project clone.
Then clean node_modules and reinstall your dependencies with e.g. `yarn`.
Then you can test it like it is a normal dependency. You need to run `pnpm run build` in your clone.

File: tests/issues/issue-317/issue-317.test.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+			printWidth: 120,
+			semi: true,
+			singleQuote: true,
+			trailingComma: 'none',
+			tabWidth: 2,
+			useTabs: false,
+
+			pugSingleQuote: false,
+			pugFramework: 'vue',
+			pugSortAttributesBeginning: [
```
Comment (Shinigami92) [inline tests/issues/issue-317/issue-317.test.ts:25]: I think you can remove this huge block. I mostly try to shrink the config for the tests down to the real problem, so only the necessary part is left over. Then you can also remove the "known-words" from eslint again, due to these are really only project specific :thinking: 
We could think about to run spellcheck only on `src`, but on the other side it's nice to have everywhere.

**Review by Shinigami92 — CHANGES_REQUESTED** [review]
```pug
v-row(no-gutters)
  v-col(cols="12")
    a.text-body-2.d-inline.mr-4(
      v-for="variable in variables",
      :key="variable",
      v-text="`$\\{${variable}}`",
      @click="copyToClipboard(variable)"
    )
```

became

```pug
v-row(no-gutters)
  v-col(cols="12")
    a(
      v-for="variable in variables",
      :key="variable",
      v-text="`$\\{${variable}}`"
    ).text-body-2.d-inline.mr-4@click="copyToClipboard(variable)"
```

😬

`@click="copyToClipboard(variable)"` should still be in the attributes

**Comment by Shinigami92** [comment]
I assume it's a problem with a regex pattern somewhere or so, because `:something` and `@something` (Vue event and expression bindings are not taken into account)

I'm a bit afraid of if this also breaks for Angular bindings with `(something)` and `[something]` syntax

**Comment by shadowgate15** [comment]
that is so weird because it works in the test.

https://github.com/prettier/plugin-pug/pull/318/files#diff-d49848e1889ef6546b7cc9120823686595638031bcf5c2392064b8b0d1437979R9

**Comment by Shinigami92** [comment]
Okay, I could try add a test later on... And I think about to release a beta version so you can test it and see all the junky jittering when using this functionally.
May have time tomorrow for that.

**Comment by shadowgate15** [comment]
jittering?

I have it setup as a linked dep on some of my pug projects and haven't seen that, since this fix. but they are all vanilla js not vue or angular.

**Comment by Shinigami92** [comment]
> jittering?
> 
> I have it setup as a linked dep on some of my pug projects and haven't seen that, since this fix. but they are all vanilla js not vue or angular.

Could you test it with an attribute that starts with an `@` or `:`?

**Comment by shadowgate15** [comment]
it's in the written test already and passes. and it worked in my projects. no jittering.

**Comment by Shinigami92** [comment]
> it's in the written test already and passes. and it worked in my projects. no jittering.

Okay, need to check it again
Hope I have time this evening

**Comment by Shinigami92** [comment]
It's my `pugSortAttributesEnd: ['^@click'],`! :eyes: 

So it has interference with the `pugSortAttributesEnd` (and I assume also with `pugSortAttributesBeginning`) option.
Could you add a further test to it? Or at least it should be sufficient to add `pugSortAttributesEnd: ['^@click'],` before the `pugClassLocation`.

File: tests/issues/issue-317/issue-317.test.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	test('should respect printWidth and pugClassLocation', () => {
+		const expected: string = readFileSync(resolve(__dirname, 'formatted.pug'), 'utf8');
+		const code: string = readFileSync(resolve(__dirname, 'unformatted.pug'), 'utf8');
+		const actual: string = format(code, {
+			parser: 'pug',
+			plugins: [plugin],
+
+			printWidth: 120,
+
+			pugClassLocation: 'after-attributes'
```
Comment (Shinigami92) [inline tests/issues/issue-317/issue-317.test.ts:16]:
```suggestion
			pugSortAttributesEnd: ['^@click'],
			pugClassLocation: 'after-attributes'
```

**Comment by shadowgate15** [comment]
yep that was the cause working on it now.

**Comment by shadowgate15** [comment]
that should do it. Let me know if you find any other edge cases.

**Comment by Shinigami92** [comment]
It works :tada: 

Only a little not-finished on first run with exact `printWidth`

After first run:

```pug
//- some other stuff
                  div(
                    v-t="'entity-management.entity-group.crud.entity-types-txt'"
                  ).text-body-2.text--secondary.mb-2.mt-4
```

After second run:

```pug
                  div(v-t="'entity-management.entity-group.crud.entity-types-txt'").text-body-2.text--secondary.mb-2.mt-4
```

But this is now 121 cols long, so exactly one of with `printWidth: 120` :thinking: 

I assume it has something to do with `this.currentLineLength`.
Would you like to tackle this also? If not and you are just satisfied, I'm also okay to merge and release it.

**Comment by shadowgate15** [comment]
I think I know where it is coming from. I’ll see if I can resolve it.

**Comment by Shinigami92** [comment]
Thank you, I will say good night for today, will read you tomorrow :)

**Comment by shadowgate15** [comment]
@Shinigami92 that should fix that problem and give an accurate count now.

**Review by Shinigami92 — APPROVED** [review]

**Comment by Shinigami92** [comment]
@shadowgate15 You can now delete your fork and directly checkout this repo.
When then creating a PR, I can more easily work together with you on that branch :slightly_smiling_face: 
You have write access, but you cannot commit directly to main branch, only custom branches/PRs

**Pull request merged**